File: server/ai/scoring.ts
import {
  CaseOpportunityScore,
  ConfidenceBreakdown,
  ScoreComponents,
  SCORING_MODEL_VERSION,
  Tier,
} from "../ontology/model";

export interface ScoreIntakeInput {
  scores: Omit<ScoreComponents, "mechanismSeverity" | "uncertaintyPenalty">;
  mechanism: string;
  evidenceCount: number;
  contradictions: {
    highSeverityUnresolved: number;
    other: number;
  };
  numParties: number;
  numResolvedParties: number;
}

export const MECHANISM_SEVERITY: Record<string, number> = {
  WheelOffIncident: 0.9,
  MechanicalFailureIncident: 0.75,
  AngleCollision: 0.65,
  SideswipeCollision: 0.6,
  DebrisIncident: 0.55,
  RearEndCollision: 0.5,
  VehicleStall: 0.3,
};

export function computeCaseOpportunityScore(s: ScoreComponents): number {
  const raw =
    0.25 * s.liability +
    0.20 * s.injury +
    0.20 * s.collectability +
    0.15 * s.evidence +
    0.10 * s.mechanismSeverity +
    0.10 * s.defendantResolution -
    0.20 * s.uncertaintyPenalty;
  return Math.max(0, Math.min(1, raw));
}

export function classifyTier(cos: number, hasUnresolvedHighSeverityContradiction: boolean): Tier {
  if (hasUnresolvedHighSeverityContradiction) return "C";
  if (cos >= 0.80) return "A";
  if (cos >= 0.65) return "B";
  if (cos >= 0.45) return "C";
  return "D";
}

export function computeUncertaintyPenalty(
  numHighSeverityUnresolved: number,
  numOther: number = 0,
): number {
  const penalty = 0.22 * numHighSeverityUnresolved + 0.05 * numOther;
  return Math.min(1, Math.round(penalty * 100) / 100);
}

export function computeEventCorrelation(evidenceCount: number): number {
  return Math.round(Math.min(1, 0.5 + 0.25 * evidenceCount) * 100) / 100;
}

export function computeCausalRelationship(
  eventCorrelation: number,
  hasAnyContradiction: boolean,
): number {
  const factor = hasAnyContradiction ? 0.6 : 0.9;
  return Math.round(eventCorrelation * factor * 100) / 100;
}

export function computePartyAttribution(numParties: number, numResolved: number): number {
  if (numParties === 0) return 0;
  return Math.round(0.5 * (numResolved / numParties) * 100) / 100;
}

export function computeConfidenceDimensions(
  evidenceCount: number,
  hasAnyContradiction: boolean,
  numParties: number,
  numResolved: number,
): ConfidenceBreakdown {
  const eventCorrelation = computeEventCorrelation(evidenceCount);
  return {
    eventCorrelation,
    causalRelationship: computeCausalRelationship(eventCorrelation, hasAnyContradiction),
    partyAttribution: computePartyAttribution(numParties, numResolved),
  };
}

export function mechanismSeverityFor(mechanism: string): number {
  return MECHANISM_SEVERITY[mechanism] ?? 0.4;
}

export function scoreIntake(input: ScoreIntakeInput): CaseOpportunityScore {
  const uncertaintyPenalty = computeUncertaintyPenalty(
    input.contradictions.highSeverityUnresolved,
    input.contradictions.other,
  );

  const scores: ScoreComponents = {
    ...input.scores,
    mechanismSeverity: mechanismSeverityFor(input.mechanism),
    uncertaintyPenalty,
  };

  const cos = computeCaseOpportunityScore(scores);
  const hasUnresolvedHighSeverity = input.contradictions.highSeverityUnresolved > 0;
  const tier = classifyTier(cos, hasUnresolvedHighSeverity);
  const confidence = computeConfidenceDimensions(
    input.evidenceCount,
    input.contradictions.highSeverityUnresolved + input.contradictions.other > 0,
    input.numParties,
    input.numResolvedParties,
  );

  const reasons: string[] = [];
  if (scores.liability > 0.5) reasons.push("Meaningful liability signal present");
  if (scores.injury > 0.3) reasons.push("Meaningful injury signal present");
  if (scores.collectability > 0.5) reasons.push("Commercial/collectable defendant likely");
  if (hasUnresolvedHighSeverity) reasons.push("Unresolved high-severity contradiction — forced to Tier C");

  return {
    score: Math.round(cos * 100) / 100,
    tier,
    reasons,
    confidence,
    modelVersion: SCORING_MODEL_VERSION,
    reviewedAt: new Date().toISOString(),
  };
}

File: app/scoring.py
from __future__ import annotations

from enum import Enum


class IncidentMechanism(str, Enum):
    REAR_END = "RearEndCollision"
    ANGLE = "AngleCollision"
    SIDESWIPE = "SideswipeCollision"
    DEBRIS = "DebrisIncident"
    WHEEL_OFF = "WheelOffIncident"
    MECHANICAL_FAILURE = "MechanicalFailureIncident"
    VEHICLE_STALL = "VehicleStall"


MECHANISM_DISPLAY_LABEL: dict[IncidentMechanism, str] = {
    IncidentMechanism.REAR_END: "Rear-End Collision",
    IncidentMechanism.ANGLE: "Angle Collision",
    IncidentMechanism.SIDESWIPE: "Sideswipe Collision",
    IncidentMechanism.DEBRIS: "Debris Incident",
    IncidentMechanism.WHEEL_OFF: "Wheel-Off Incident",
    IncidentMechanism.MECHANICAL_FAILURE: "Mechanical Failure",
    IncidentMechanism.VEHICLE_STALL: "Vehicle Stall",
}

MECHANISM_SEVERITY: dict[IncidentMechanism, float] = {
    IncidentMechanism.WHEEL_OFF: 0.90,
    IncidentMechanism.MECHANICAL_FAILURE: 0.75,
    IncidentMechanism.ANGLE: 0.65,
    IncidentMechanism.SIDESWIPE: 0.60,
    IncidentMechanism.DEBRIS: 0.55,
    IncidentMechanism.REAR_END: 0.50,
    IncidentMechanism.VEHICLE_STALL: 0.30,
}


def compute_case_opportunity_score(
    liability: float,
    injury: float,
    collectability: float,
    evidence: float,
    mechanism_severity: float,
    defendant_resolution: float,
    uncertainty_penalty: float,
) -> float:
    """Canonical G-CCI Case Opportunity Score.

    COS = .25L + .20I + .20C + .15E + .10M + .10D - .20U
    """
    raw = (
        0.25 * liability
        + 0.20 * injury
        + 0.20 * collectability
        + 0.15 * evidence
        + 0.10 * mechanism_severity
        + 0.10 * defendant_resolution
        - 0.20 * uncertainty_penalty
    )
    return max(0.0, min(1.0, raw))


def classify_tier(cos: float, has_unresolved_high_severity_contradiction: bool) -> str:
    if has_unresolved_high_severity_contradiction:
        return "C"
    if cos >= 0.80:
        return "A"
    if cos >= 0.65:
        return "B"
    if cos >= 0.45:
        return "C"
    return "D"


def compute_uncertainty_penalty(num_high_severity_unresolved: int, num_other: int = 0) -> float:
    return min(1.0, round(0.22 * num_high_severity_unresolved + 0.05 * num_other, 2))


def compute_event_correlation(evidence_count: int) -> float:
    return round(min(1.0, 0.5 + 0.25 * evidence_count), 2)


def compute_causal_relationship(event_correlation: float, has_any_contradiction: bool) -> float:
    return round(event_correlation * (0.9 if not has_any_contradiction else 0.6), 2)


def compute_party_attribution(num_parties: int, num_resolved: int) -> float:
    return round(0.5 * (num_resolved / num_parties), 2) if num_parties else 0.0


EXPECTED_EVIDENCE_TYPES = ["Crash Report", "CCTV", "CAD Record", "Tow Record", "EDR", "ELD"]
